karpenter-operator/controllers/nodeclass/ec2_nodeclass_controller.go

```diff
@@ -0,0 +1,394 @@
+					"openshiftec2nodeclasses.karpenter.hypershift.openshift.io":
+					return []ctrl.Request{{NamespacedName: client.ObjectKey{Namespace: r.Namespace}}}
+				}
+				return nil
+			},
+		)).
+		Watches(&awskarpenterv1.EC2NodeClass{}, &handler.EnqueueRequestForObject{})
+
+	return bldr.Complete(r)
+}
+
+func (r *EC2NodeClassReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+	log.Info("Reconciling", "req", req)
+
+	hcp, err := r.getHCP(ctx)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	if err := r.reconcileCRDs(ctx, false); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	openshiftEC2NodeClass := &hyperkarpenterv1.OpenshiftEC2NodeClass{}
+	if err := r.guestClient.Get(ctx, req.NamespacedName, openshiftEC2NodeClass); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info("openshiftEC2NodeClass not found, aborting reconcile", "name", req.NamespacedName)
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, fmt.Errorf("failed to get openshiftEC2NodeClass %q: %w", req.NamespacedName, err)
+	}
+
+	ec2NodeClass := &awskarpenterv1.EC2NodeClass{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      openshiftEC2NodeClass.Name,
+			Namespace: openshiftEC2NodeClass.Namespace,
+		},
+	}
+
+	if !openshiftEC2NodeClass.DeletionTimestamp.IsZero() {
+		exists, err := util.DeleteIfNeeded(ctx, r.guestClient, ec2NodeClass)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		if exists {
+			// wait until EC2NodeClass is deleted
+			return ctrl.Result{RequeueAfter: time.Second * 5}, nil
+		}
+
+		if controllerutil.ContainsFinalizer(openshiftEC2NodeClass, finalizer) {
+			original := openshiftEC2NodeClass.DeepCopy()
+			controllerutil.RemoveFinalizer(openshiftEC2NodeClass, finalizer)
+			if err := r.guestClient.Patch(ctx, openshiftEC2NodeClass, client.MergeFromWithOptions(original, client.MergeFromWithOptimisticLock{})); err != nil {
+				return ctrl.Result{}, fmt.Errorf("failed to remove finalizer from openshiftEC2NodeClass: %w", err)
+			}
+		}
+		return ctrl.Result{}, nil
+	}
+
+	if !controllerutil.ContainsFinalizer(hcp, finalizer) {
+		original := openshiftEC2NodeClass.DeepCopy()
+		controllerutil.AddFinalizer(openshiftEC2NodeClass, finalizer)
+		if err := r.guestClient.Patch(ctx, openshiftEC2NodeClass, client.MergeFromWithOptions(original, client.MergeFromWithOptimisticLock{})); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to add finalizer to openshiftEC2NodeClass: %w", err)
+		}
+	}
+
+	userDataSecret, err := r.getUserDataSecret(ctx, hcp)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if _, err := r.CreateOrUpdate(ctx, r.guestClient, ec2NodeClass, func() error {
+		return reconcileEC2NodeClass(ec2NodeClass, openshiftEC2NodeClass, hcp, userDataSecret)
+	}); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if err := r.reconcileStatus(ctx, ec2NodeClass, openshiftEC2NodeClass); err != nil {
```

Comment:
why are we not following the same CreateOrUpdate pattern here?

Reply:
CreateOrUpdate  doesn't update status right? we would need to add the type explicitly in upsert.go to for it to work. I just followed what we do for HCP with the client patch.